File: linked-lists/09-reverse-linked-list.py
"""
Problem 09 (Bonus): Reverse Linked List (LeetCode #206)
Topic: Linked Lists
Difficulty: Easy-Medium

Description:
Given the head of a singly linked list, reverse the list, and return the reversed list.

Time Complexity: O(n)
Space Complexity: O(1)
"""

from typing import Optional

class ListNode:
    def __init__(self, val: int = 0, next: Optional['ListNode'] = None):
        self.val = val
        self.next = next

class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        prev = None
        curr = head
        while curr:
            next_node = curr.next
            curr.next = prev
            prev = curr
            curr = next_node
        return prev

def create_linked_list(arr):
    dummy = ListNode(0)
    curr = dummy
    for val in arr:
        curr.next = ListNode(val)
        curr = curr.next
    return dummy.next

def print_linked_list(head):
    vals = []
    curr = head
    while curr:
        vals.append(curr.val)
        curr = curr.next
    return vals

if __name__ == "__main__":
    sol = Solution()
    head1 = create_linked_list([1, 2, 3, 4, 5])
    rev1 = sol.reverseList(head1)
    print("Test 1:", print_linked_list(rev1))  # Expected: [5, 4, 3, 2, 1]

    head2 = create_linked_list([1, 2])
    rev2 = sol.reverseList(head2)
    print("Test 2:", print_linked_list(rev2))  # Expected: [2, 1]

File: linked-lists/09-reverse-linked-list.c
/*
 * Problem 09 (Bonus): Reverse Linked List (LeetCode #206)
 * Topic: Linked Lists
 * Difficulty: Easy-Medium
 * Language: C
 */

#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* reverseList(struct ListNode* head) {
    struct ListNode* prev = NULL;
    struct ListNode* curr = head;
    while (curr != NULL) {
        struct ListNode* nextTemp = curr->next;
        curr->next = prev;
        prev = curr;
        curr = nextTemp;
    }
    return prev;
}

struct ListNode* createNode(int val) {
    struct ListNode* node = (struct ListNode*)malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = NULL;
    return node;
}

int main() {
    struct ListNode* head = createNode(1);
    head->next = createNode(2);
    head->next->next = createNode(3);
    
    struct ListNode* rev = reverseList(head);
    printf("Test 1: ");
    struct ListNode* curr = rev;
    while (curr) {
        printf("%d ", curr->val);
        curr = curr->next;
    }
    printf("\n"); // Expected: 3 2 1
    return 0;
}
